File: Excel/548 Challenge.py
import pandas as pd
from itertools import chain

path = "548 Tap Code Cipher.xlsx"
input1 = pd.read_excel(path, sheet_name='Sheet1', usecols="A:F", nrows=6)
input2 = pd.read_excel(path, sheet_name='Sheet1', usecols="H:H", nrows=10)
test = pd.read_excel(path, sheet_name='Sheet1', usecols="I:I", nrows=10)

coding_table = input1.melt(id_vars=input1.columns[0], var_name='letter', value_name='code')
coding_table.columns = ['row', 'col', 'letter']
coding_table = coding_table.assign(letter=coding_table['letter'].str.lower().str.split('/')).explode('letter').dropna().reset_index(drop=True)
coding_table['row'] = coding_table['row'].astype(int)

def encrypt(word):
    return ' '.join(
        ['.' * int(num) for num in ' '.join(
            [f"{coding_table.loc[coding_table['letter'] == char, 'row'].values[0]} {coding_table.loc[coding_table['letter'] == char, 'col'].values[0]}"
             for char in word]
        ).split()]
    )

input2['Answer Expected'] = input2['Words'].apply(encrypt)
print(input2['Answer Expected'].equals(test['Answer Expected'])) # True

File: Excel/500-599/548/548 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/548 Tap Code Cipher.xlsx"
input1 = read_excel(path, range = "A1:F6")
input2 = read_excel(path, range = "H1:H10")
test  = read_excel(path, range = "I1:I10")

coding_table = input1 %>%
  pivot_longer(-1, names_to = "letter", values_to = "code") %>%
  select(row = 1, col = 2, letter = 3) %>%
  separate_rows(letter, sep = "/") %>%
  mutate(letter = str_to_lower(letter))

encrypt  = function(word) {
  characters = str_split(word, "") %>% unlist()

  coord = map_dfr(characters, ~{
    row = coding_table %>% filter(letter == .x) %>% pull(row)
    col = coding_table %>% filter(letter == .x) %>% pull(col)
    tibble(row = row, col = col) 
  }) %>%
    unite("coord", row, col, sep = " ") %>%
    pull(coord) %>%
    paste(collapse = " ")
  
  coord = str_split(coord, " ") %>% unlist() %>%
    map_dfr(~{
      dots = str_c(rep(".", .x), collapse = "")
      tibble(dots = dots)
    }) %>%
    pull(dots) %>%
    paste(collapse = " ")
    
  return(coord)
}

result = input2 %>%
  mutate(`Answer Expected` = map_chr(Words, encrypt))

identical(test$`Answer Expected`, result$`Answer Expected`)
#> [1] TRUE